test/common/utxobased/db/NewProcessor.spec.ts

```diff
@@ -0,0 +1,322 @@
+    expect(tx1.ourOuts[0]).to.eqls('0')
+    expect(tx1.ourAmount).to.eqls('1')
+    const [txByBlockHeight1] = await processor.fetchTransactions({
+      blockHeight: 1
+    })
+    expect(txByBlockHeight1.blockHeight).to.eqls(1)
+
+    // insert the same transaction, but with a script pubkey referencing an input
+    await processor.insertTransaction({
+      tx: transaction1,
+      scriptPubkey: 'pubkeyin1'
+    })
+
+    const [tx2] = await processor.fetchTransactions({ txId: 'transaction1' })
+    expect(tx2.ourOuts[0]).to.eqls('0')
+    expect(tx2.ourIns[0]).to.eqls('0')
+    expect(tx2.ourAmount).to.eqls('0')
+
+    // insert the same transaction, but with a script pubkey referencing another output
+    await processor.insertTransaction({
+      tx: transaction1,
+      scriptPubkey: 'pubkeyout2'
+    })
+
+    const [tx3] = await processor.fetchTransactions({ txId: 'transaction1' })
+    expect(tx3.ourOuts[1]).to.eqls('1')
+    expect(tx3.ourIns[0]).to.eqls('0')
+    expect(tx3.ourAmount).to.eqls('1')
+
+    const [tx4] = await processor.fetchTransactions({ options: {} })
+    expect(tx4).not.to.be.undefined
+
+    const [tx5] = await processor.fetchTransactions({
+      options: {
+        startDate: new Date(11),
+        endDate: new Date(15)
+      }
+    })
+    expect(tx5).to.be.undefined
+
+    const [tx6] = await processor.fetchTransactions({
+      options: {
+        startDate: new Date(9),
+        endDate: new Date(15)
+      }
+    })
+    expect(tx6).not.to.be.undefined
+  })
+
+  it('insert multiple transactions to baselets', async () => {
+    const storage = {}
+    const disklet = makeMemoryDisklet(storage)
+    processor = await makeNewProcessor({ disklet })
+
+    const input1: ITransactionInput = {
+      txId: 'random',
+      outputIndex: 0,
+      scriptPubkey: 'pubkeyin1',
+      n: 0,
+      amount: '1'
+    }
+    const output1: ITransactionOutput = {
+      amount: '1',
+      n: 0,
+      scriptPubkey: 'pubkeyout1'
+    }
+    const output2: ITransactionOutput = {
+      amount: '1',
+      n: 1,
+      scriptPubkey: 'pubkeyout2'
+    }
+    const transaction1: IProcessorTransaction = {
+      txid: 'transaction1',
+      hex: '',
+      blockHeight: 1,
+      date: new Date(10).getTime(),
+      fees: '1',
+      inputs: [input1],
+      outputs: [output1, output2],
+      ourIns: [],
+      ourOuts: [],
+      ourAmount: '0'
+    }
+
+    const transaction2: IProcessorTransaction = {
+      txid: 'transaction2',
+      hex: '',
+      blockHeight: 1,
+      date: new Date(20).getTime(),
+      fees: '1',
+      inputs: [input1],
+      outputs: [output1, output2],
+      ourIns: [],
+      ourOuts: [],
+      ourAmount: '0'
+    }
+
+    await processor.insertTransaction({
+      tx: transaction1,
+      scriptPubkey: 'pubkeyout1'
+    })
+    await processor.insertTransaction({
+      tx: transaction2,
+      scriptPubkey: 'pubkeyout2'
+    })
```

Comment:
I'm starting to understand what `ourIns` and `ourOuts` is. However, I'm a bit confused about why we track input/output ownership this way. You can see how you have two different transaction objects with identical inputs and outputs which is invalid from a consensus point of view. I'm aware this is dummy txs for testing, however shouldn't the processors check for conflicting txs like this?

Even if we used the the same transaction with a different `scriptPubkey` for each `insertTransaction` call, wouldn't we still be adding the same transaction into our data layer twice or would the last insert stomp on the previous? Would it make sense to let `scriptPubkey` be `scriptPubkeys: string[]` and allow for one or more; this way we can declare ownership of more than one `output` in the transaction (which is very possible for a transaction).

I may be misunderstanding things here, so I'm happy to discuss more.

Reply:
> You can see how you have two different transaction objects with identical inputs and outputs which is invalid from a consensus point of view. I'm aware this is dummy txs for testing, however shouldn't the processors check for conflicting txs like this?

It's perfectly valid from a consensus point of view that transactions can have identical outputs. The input in our serialization here also contains the previous (now consumed output's) transaction id. This txid needn't be unique either! See this stackexchange question: https://bitcoin.stackexchange.com/questions/75300/what-would-happened-if-two-transactions-have-the-same-hash . Check how the coinbase transactions here: https://www.blockchain.com/btc/block/91812 and here: https://www.blockchain.com/btc/block/91842 have the same txid! Nowadays it's super uncommon for this to happen, since there are policy rules in place making it harder, but it's not cryptographically impossible.

> Even if we used the the same transaction with a different scriptPubkey for each insertTransaction call, wouldn't we still be adding the same transaction into our data layer twice or would the last insert stomp on the previous? 

The last insert stomps the previous, updating ourIns and ourOuts based on the passed in scriptPubkey.

> Would it make sense to let scriptPubkey be scriptPubkeys: string[] and allow for one or more; this way we can declare ownership of more than one output in the transaction (which is very possible for a transaction).

Not really. The network layer passes in one or more transactions for a single scriptPubkey at a time. This means for a transaction actually containing one ourIn and one ourOut, we insert it twice. The first time with the scriptPubkey of the input populating ourIn, the second time with the scriptPubkey of the output populating ourOut.